File: internal/handlers/chat_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"
	"chat-backend/internal/websocket"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ChatHandler struct {
	db  *database.Database
	hub *websocket.Hub
}

func NewChatHandler(db *database.Database, hub *websocket.Hub) *ChatHandler {
	return &ChatHandler{db: db, hub: hub}
}

type CreateChatRequest struct {
	Type        string   `json:"type" binding:"required"` // direct, group
	MemberIDs   []string `json:"member_ids"`
	GroupName   string   `json:"group_name,omitempty"`
	IsAnonymous bool     `json:"is_anonymous"` // current user's identity hidden from others
}

// SendMessage is now handled by MessageHandler

func (h *ChatHandler) GetChats(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	cursor, err := h.db.MongoDB.Collection("chats").Find(
		context.Background(),
		bson.M{"members": userIDObj},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch chats"})
		return
	}
	defer cursor.Close(context.Background())

	var chats []models.Chat
	if err := cursor.All(context.Background(), &chats); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode chats"})
		return
	}

	// Enrich: for direct chats, set other_party_anonymous when the other member is anonymous to current user
	out := make([]map[string]interface{}, 0, len(chats))
	for _, ch := range chats {
		m := map[string]interface{}{
			"id":         ch.ID,
			"type":       ch.Type,
			"members":    ch.Members,
			"group_name": ch.GroupName,
			"created_at": ch.CreatedAt,
			"updated_at": ch.UpdatedAt,
		}
		if ch.Type == "direct" && ch.AnonymousFromUserID != nil && len(ch.Members) == 2 {
			// The other party is anonymous to me if AnonymousFromUserID is the other user (not me)
			otherIsAnonymous := *ch.AnonymousFromUserID != userIDObj
			m["other_party_anonymous"] = otherIsAnonymous
		}
		out = append(out, m)
	}
	c.JSON(http.StatusOK, out)
}

func (h *ChatHandler) CreateChat(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req CreateChatRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	members := []primitive.ObjectID{userIDObj}
	for _, memberIDStr := range req.MemberIDs {
		memberID, err := primitive.ObjectIDFromHex(memberIDStr)
		if err != nil {
			continue
		}
		members = append(members, memberID)
	}

	var anonymousFrom *primitive.ObjectID
	if req.IsAnonymous && req.Type == "direct" {
		anonymousFrom = &userIDObj
	}

	chat := models.Chat{
		ID:                  primitive.NewObjectID(),
		Type:                req.Type,
		Members:             members,
		GroupName:           req.GroupName,
		AnonymousFromUserID: anonymousFrom,
		CreatedAt:           time.Now(),
		UpdatedAt:           time.Now(),
	}

	_, err := h.db.MongoDB.Collection("chats").InsertOne(context.Background(), chat)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create chat"})
		return
	}

	c.JSON(http.StatusCreated, chat)
}

func (h *ChatHandler) GetChat(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	chatIDStr := c.Param("chat_id")
	chatID, err := primitive.ObjectIDFromHex(chatIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid chat ID"})
		return
	}

	var chat models.Chat
	err = h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{"_id": chatID},
	).Decode(&chat)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Chat not found"})
		return
	}

	out := map[string]interface{}{
		"id":         chat.ID,
		"type":       chat.Type,
		"members":    chat.Members,
		"group_name": chat.GroupName,
		"created_at": chat.CreatedAt,
		"updated_at": chat.UpdatedAt,
	}
	if chat.Type == "direct" && chat.AnonymousFromUserID != nil && len(chat.Members) == 2 {
		out["other_party_anonymous"] = *chat.AnonymousFromUserID != userIDObj
	}
	c.JSON(http.StatusOK, out)
}

func (h *ChatHandler) GetMessages(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	chatIDStr := c.Param("chat_id")
	chatID, err := primitive.ObjectIDFromHex(chatIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid chat ID"})
		return
	}

	var chat models.Chat
	if err := h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{"_id": chatID},
	).Decode(&chat); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Chat not found"})
		return
	}

	cursor, err := h.db.MongoDB.Collection("messages").Find(
		context.Background(),
		bson.M{
			"chat_id":     chatID,
			"is_deleted":  false,
			"deleted_for": bson.M{"$ne": userIDObj},
		},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch messages"})
		return
	}
	defer cursor.Close(context.Background())

	var messages []models.Message
	if err := cursor.All(context.Background(), &messages); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode messages"})
		return
	}

	// Mask anonymous sender for the recipient (do not expose sender_id when chat has AnonymousFromUserID)
	var out []map[string]interface{}
	for _, m := range messages {
		raw := map[string]interface{}{
			"id":           m.ID,
			"chat_id":      m.ChatID,
			"sender_id":    m.SenderID,
			"content":      m.Content,
			"message_type": m.MessageType,
			"is_anonymous": m.IsAnonymous,
			"status":       m.Status,
			"created_at":   m.CreatedAt,
			"updated_at":   m.UpdatedAt,
		}
		if chat.AnonymousFromUserID != nil && m.SenderID == *chat.AnonymousFromUserID && m.SenderID != userIDObj {
			raw["sender_id"] = nil
			raw["is_anonymous"] = true
		}
		out = append(out, raw)
	}
	c.JSON(http.StatusOK, out)
}

// SendMessage is now handled by MessageHandler.SendMessage
// This endpoint redirects to /messages endpoint
func (h *ChatHandler) SendMessage(c *gin.Context) {
	// Redirect to message handler
	messageHandler := NewMessageHandler(h.db, h.hub)
	messageHandler.SendMessage(c)
}

func (h *ChatHandler) DeleteMessage(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	messageIDStr := c.Param("message_id")
	messageID, err := primitive.ObjectIDFromHex(messageIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid message ID"})
		return
	}

	// Verify message belongs to user
	var message models.Message
	err = h.db.MongoDB.Collection("messages").FindOne(
		context.Background(),
		bson.M{"_id": messageID, "sender_id": userIDObj},
	).Decode(&message)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Message not found"})
		return
	}

	// Soft delete
	_, err = h.db.MongoDB.Collection("messages").UpdateOne(
		context.Background(),
		bson.M{"_id": messageID},
		bson.M{"$set": bson.M{
			"is_deleted": true,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete message"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Message deleted successfully"})
}

func (h *ChatHandler) DeleteChat(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	chatIDStr := c.Param("chat_id")
	chatID, err := primitive.ObjectIDFromHex(chatIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid chat ID"})
		return
	}

	// Verify user is a member of the chat
	var chat models.Chat
	err = h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{"_id": chatID, "members": userIDObj},
	).Decode(&chat)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Chat not found or you are not a member"})
		return
	}

	// Remove user from members list (soft delete - chat will be hidden from user)
	// This works for both direct and group chats
	_, err = h.db.MongoDB.Collection("chats").UpdateOne(
		context.Background(),
		bson.M{"_id": chatID},
		bson.M{"$pull": bson.M{"members": userIDObj}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete chat"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Chat deleted successfully"})
}

File: src/handlers/chatHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { randomUUID } = require('crypto');
const { getDB } = require('../database');
const { hub } = require('../websocket/hub');

// ─── Helpers ──────────────────────────────────────────────────────────────────

function maskAnonymousSender(message, chatAnonymousFromUserId, userId) {
  if (!chatAnonymousFromUserId) return message;
  // The anonymous user is the one who set anonymous_from_user_id.
  // For the OTHER party (not the sender), mask sender info.
  const senderId = message.sender_id ? message.sender_id.toString() : null;
  const anonFromId = chatAnonymousFromUserId.toString();
  // If the sender is the anonymous user and the viewer is not the sender, mask
  if (senderId === anonFromId && userId !== anonFromId) {
    return { ...message, sender_id: null, is_anonymous: true };
  }
  return message;
}

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function getChats(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;

    const chats = await db.collection('chats').find({
      members: userId,
    }).toArray();

    const result = [];

    for (const chat of chats) {
      const chatObj = {
        id: chat._id.toString(),
        type: chat.type,
        members: chat.members,
        group_name: chat.group_name,
        created_at: chat.created_at,
        updated_at: chat.updated_at,
      };

      // Anonymous flag for direct chats
      if (
        chat.type === 'direct' &&
        chat.anonymous_from_user_id &&
        Array.isArray(chat.members) &&
        chat.members.length === 2
      ) {
        chatObj.other_party_anonymous = chat.anonymous_from_user_id.toString() !== userId;
      }

      // Last message
      if (chat.last_message_id) {
        try {
          const lastMsg = await db.collection('messages').findOne({
            _id: new ObjectId(chat.last_message_id),
            is_deleted: { $ne: true },
            deleted_for: { $nin: [userId] },
          });

          if (lastMsg) {
            let msgObj = {
              id: lastMsg._id.toString(),
              content: lastMsg.content,
              message_type: lastMsg.message_type,
              sender_id: lastMsg.sender_id,
              is_anonymous: lastMsg.is_anonymous || false,
              status: lastMsg.status,
              created_at: lastMsg.created_at,
            };

            // Mask anonymous sender
            if (chat.anonymous_from_user_id) {
              const senderId = lastMsg.sender_id ? lastMsg.sender_id.toString() : null;
              const anonFromId = chat.anonymous_from_user_id.toString();
              if (senderId === anonFromId && userId !== anonFromId) {
                msgObj.sender_id = null;
                msgObj.is_anonymous = true;
              }
            }

            chatObj.last_message = msgObj;
          }
        } catch (_) {
          // ignore invalid last_message_id
        }
      }

      // Unread count
      if (chat.unread_count && chat.unread_count[userId] !== undefined) {
        chatObj.unread_count = chat.unread_count[userId];
      }

      result.push(chatObj);
    }

    return res.json(result);
  } catch (err) {
    console.error('getChats error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function createChat(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { type, member_ids, group_name, is_anonymous } = req.body;

    const members = [userId];
    if (Array.isArray(member_ids)) {
      for (const mid of member_ids) {
        if (!members.includes(mid)) members.push(mid);
      }
    }

    const now = new Date();
    const chatDoc = {
      type: type || 'direct',
      members,
      group_name: group_name || null,
      created_at: now,
      updated_at: now,
    };

    if (is_anonymous && (type === 'direct' || !type)) {
      chatDoc.anonymous_from_user_id = userId;
    }

    const insertResult = await db.collection('chats').insertOne(chatDoc);
    chatDoc._id = insertResult.insertedId;
    const chatIdStr = insertResult.insertedId.toString();
    const now2 = chatDoc.created_at;

    // Notify every member EXCEPT the creator so their sidebar updates in real-time
    for (const memberId of members) {
      if (memberId !== userId) {
        hub.sendToUser(memberId.toString(), {
          type: 'new_chat',
          chat_id: chatIdStr,
          chat_type: chatDoc.type,
          group_name: chatDoc.group_name || null,
          members,
          created_by: userId,
          message_id: randomUUID(),
          timestamp: now2.toISOString(),
        });
      }
    }

    return res.status(201).json(chatDoc);
  } catch (err) {
    console.error('createChat error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getChat(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const chatIdStr = req.params.chat_id;

    let chatObjId;
    try {
      chatObjId = new ObjectId(chatIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid chat ID' });
    }

    const chat = await db.collection('chats').findOne({ _id: chatObjId });
    if (!chat) return res.status(404).json({ error: 'Chat not found' });

    const chatObj = {
      id: chat._id.toString(),
      type: chat.type,
      members: chat.members,
      group_name: chat.group_name,
      created_at: chat.created_at,
      updated_at: chat.updated_at,
    };

    if (
      chat.type === 'direct' &&
      chat.anonymous_from_user_id &&
      Array.isArray(chat.members) &&
      chat.members.length === 2
    ) {
      chatObj.other_party_anonymous = chat.anonymous_from_user_id.toString() !== userId;
    }

    return res.json(chatObj);
  } catch (err) {
    console.error('getChat error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getMessages(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const chatIdStr = req.params.chat_id;

    let chatObjId;
    try {
      chatObjId = new ObjectId(chatIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid chat ID' });
    }

    // Verify chat exists and user is member
    const chat = await db.collection('chats').findOne({ _id: chatObjId });
    if (!chat) return res.status(404).json({ error: 'Chat not found' });

    const messages = await db.collection('messages').find({
      chat_id: chatIdStr,
      is_deleted: { $ne: true },
      deleted_for: { $nin: [userId] },
    }).sort({ created_at: 1 }).toArray();

    const result = [];

    for (const msg of messages) {
      let msgObj = {
        id: msg._id.toString(),
        chat_id: msg.chat_id,
        content: msg.content,
        message_type: msg.message_type,
        sender_id: msg.sender_id,
        is_anonymous: msg.is_anonymous || false,
        status: msg.status,
        created_at: msg.created_at,
        updated_at: msg.updated_at,
        is_edited: msg.is_edited || false,
        reactions: msg.reactions || [],
        file_url: msg.file_url,
        file_name: msg.file_name,
        file_size: msg.file_size,
        thumbnail_url: msg.thumbnail_url,
        duration: msg.duration,
        is_pinned: msg.is_pinned || false,
      };

      // Mask anonymous sender
      if (chat.anonymous_from_user_id) {
        const senderId = msg.sender_id ? msg.sender_id.toString() : null;
        const anonFromId = chat.anonymous_from_user_id.toString();
        if (senderId === anonFromId && userId !== anonFromId) {
          msgObj.sender_id = null;
          msgObj.is_anonymous = true;
        }
      }

      // Populate reply_to
      if (msg.reply_to_id) {
        try {
          const replyMsg = await db.collection('messages').findOne({
            _id: new ObjectId(msg.reply_to_id),
          });
          if (replyMsg) {
            msgObj.reply_to = {
              id: replyMsg._id.toString(),
              content: replyMsg.content,
              message_type: replyMsg.message_type,
              sender_id: replyMsg.sender_id,
            };
          }
        } catch (_) {
          // ignore
        }
      }

      result.push(msgObj);
    }

    return res.json(result);
  } catch (err) {
    console.error('getMessages error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function sendMessage(req, res) {
  // Delegate to messageHandler to avoid code duplication
  // Override the chat_id param to come from req.params.chat_id
  const messageHandler = require('./messageHandler');
  return messageHandler.sendMessage(req, res);
}

async function deleteMessage(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const messageIdStr = req.params.message_id;
    const { delete_for_everyone } = req.body;

    let messageObjId;
    try {
      messageObjId = new ObjectId(messageIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid message ID' });
    }

    const message = await db.collection('messages').findOne({ _id: messageObjId });
    if (!message) return res.status(404).json({ error: 'Message not found' });

    if (message.sender_id !== userId) {
      return res.status(403).json({ error: 'Forbidden' });
    }

    if (delete_for_everyone) {
      await db.collection('messages').updateOne(
        { _id: messageObjId },
        { $set: { is_deleted: true, updated_at: new Date() } }
      );
    } else {
      await db.collection('messages').updateOne(
        { _id: messageObjId },
        { $addToSet: { deleted_for: userId }, $set: { updated_at: new Date() } }
      );
    }

    return res.json({ message: 'Message deleted' });
  } catch (err) {
    console.error('deleteMessage error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteChat(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const chatIdStr = req.params.chat_id;

    let chatObjId;
    try {
      chatObjId = new ObjectId(chatIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid chat ID' });
    }

    const chat = await db.collection('chats').findOne({ _id: chatObjId });
    if (!chat) return res.status(404).json({ error: 'Chat not found' });

    if (!Array.isArray(chat.members) || !chat.members.includes(userId)) {
      return res.status(403).json({ error: 'Forbidden' });
    }

    const now = new Date();

    // Pull userId from members
    await db.collection('chats').updateOne(
      { _id: chatObjId },
      { $pull: { members: userId }, $set: { updated_at: now } }
    );

    // Notify remaining members so their sidebar removes/updates this chat
    const remaining = chat.members.filter(m => m.toString() !== userId);
    for (const memberId of remaining) {
      hub.sendToUser(memberId.toString(), {
        type: 'chat_updated',
        chat_id: chatIdStr,
        event: 'member_left',
        left_user_id: userId,
        message_id: randomUUID(),
        timestamp: now.toISOString(),
      });
    }

    return res.json({ message: 'Left chat' });
  } catch (err) {
    console.error('deleteChat error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  getChats,
  createChat,
  getChat,
  getMessages,
  sendMessage,
  deleteMessage,
  deleteChat,
};
